test/e2e/federated-service.go

```diff
@@ -0,0 +1,283 @@
+		}
+		for _, name := range svcDNSNames {
+			discoverService(f, name, true)
+		}
+	})
+})
+
+// waitForFederatedServiceShard waits until the number of shards of a given federated
+// service reaches the expected value, i.e. numSvcs in the given individual Kubernetes
+// cluster. If the shard count, i.e. numSvcs is expected to be at least one, then
+// it also checks if the first shard's name and spec matches that of the given service.
+func waitForFederatedServiceShard(cs *release_1_3.Clientset, namespace string, service *api.Service, numSvcs int) {
+	By("Fetching a federated service shard")
+	var clSvcList *v1.ServiceList
+	if err := wait.PollImmediate(framework.Poll, FederatedServiceTimeout, func() (bool, error) {
+		var err error
+		clSvcList, err = cs.Core().Services(namespace).List(api.ListOptions{})
+		if err != nil {
+			return false, err
+		}
+		n := len(clSvcList.Items)
+		if n == numSvcs {
+			return true, nil
+		}
+		framework.Logf("%d services found, waiting for %d, trying again in %s", n, numSvcs, framework.Poll)
+		return false, nil
+	}); err != nil {
+		framework.Failf("Failed to list registered clusters: %+v", err)
+	}
+
+	if numSvcs > 0 && service != nil {
+		// Renaming for clarity/readability
+		clSvc := clSvcList.Items[0]
+		Expect(clSvc.Name).To(Equal(service.Name))
+		Expect(clSvc.Spec).To(Equal(service.Spec))
+	}
+}
+
+func createService(fcs *federation_internalclientset.Clientset, clusterClientSets []*release_1_3.Clientset, namespace string) {
+	By("Creating a federated service")
+	labels := map[string]string{
+		"foo": "bar",
+	}
+
+	svc1port := "svc1"
+	svc2port := "svc2"
+
+	service := &api.Service{
+		ObjectMeta: api.ObjectMeta{
+			Name: FederatedServiceName,
+		},
+		Spec: api.ServiceSpec{
+			Selector: labels,
+			Ports: []api.ServicePort{
+				{
+					Name:       "portname1",
+					Port:       80,
+					TargetPort: intstr.FromString(svc1port),
+				},
+				{
+					Name:       "portname2",
+					Port:       81,
+					TargetPort: intstr.FromString(svc2port),
+				},
+			},
+		},
+	}
+	_, err := fcs.Core().Services(namespace).Create(service)
+	Expect(err).NotTo(HaveOccurred())
+	for _, cs := range clusterClientSets {
+		waitForFederatedServiceShard(cs, namespace, service, 1)
+	}
+}
+
+func discoverService(f *framework.Framework, name string, exists bool) {
+	pod := &api.Pod{
+		ObjectMeta: api.ObjectMeta{
+			Name:   FederatedServicePod,
+			Labels: map[string]string{"name": FederatedServicePod},
+		},
+		Spec: api.PodSpec{
+			Containers: []api.Container{
+				{
+					Name:    "federated-service-discovery-container",
+					Image:   "gcr.io/google_containers/busybox:1.24",
+					Command: []string{"sh", "-c", "nslookup", name},
+				},
+			},
+			RestartPolicy: api.RestartPolicyNever,
+		},
+	}
+	if exists {
+		f.TestContainerOutputRegexp("federated service discovery", pod, 0, []string{
+			`Name:\s+` + FederatedDNS1123Regexp + `\nAddress \d+:\s+` + FederatedIPAddrRegexp + `\s+` + FederatedDNS1123Regexp,
```

Comment:
nit: Surely you want the actual name of the service in FederatedDNS1123Regexp?  Otherwise you'll match any DNS name that comes back.  I guess that doesn't really matter in practise.  To make this test more robust I think it would be better to only distinguish between DNS lookup success and failure.  The exact output and format of the nslookup tool is mostly irrelevant, and might change over time and/or platform.  I checked the nslookup tool and unfortunately it returns zero on both success and failure.  

But the 'host' tool is a little better for this.  It is pretty ubiquitous, and prints the name and address on success, but prints nothing on failure. That might be a simpler, more robust test.

```
$ host www.google.com
www.google.com has address 172.217.2.228
$ host rubbish
$
```